File: 大一上/图灵班/程序设计荣誉课程-1/yoj全解/YOJ-87/yoj-87.py
'''
Author: zmlad
Date: 2025-11-27 18:50:02
'''
import math
import sys

def main():
    # 读取输入
    data = sys.stdin.read().splitlines()
    n, m = map(int, data[0].split())
    
    # 解析图片数据
    image = []
    for i in range(1, n + 1):
        tokens = data[i].split()
        row = []
        for j in range(m):
            # 每个点占用6个token: R, r_val, G, g_val, B, b_val
            idx = j * 6
            r_val = int(tokens[idx + 1])
            g_val = int(tokens[idx + 3])
            b_val = int(tokens[idx + 5])
            row.append([r_val, g_val, b_val])
        image.append(row)
    
    # 修复过程
    changed = True
    while changed:
        changed = False
        # 深拷贝当前图片状态
        temp = [[[p[0], p[1], p[2]] for p in row] for row in image]
        
        for i in range(n):
            for j in range(m):
                # 如果是坏点
                if image[i][j][0] == 0 and image[i][j][1] == 0 and image[i][j][2] == 0:
                    sum_r, sum_g, sum_b = 0, 0, 0
                    count = 0
                    
                    # 检查四个方向的相邻点
                    if i > 0 and not (image[i-1][j][0] == 0 and image[i-1][j][1] == 0 and image[i-1][j][2] == 0):
                        sum_r += image[i-1][j][0]
                        sum_g += image[i-1][j][1]
                        sum_b += image[i-1][j][2]
                        count += 1
                    
                    if i < n-1 and not (image[i+1][j][0] == 0 and image[i+1][j][1] == 0 and image[i+1][j][2] == 0):
                        sum_r += image[i+1][j][0]
                        sum_g += image[i+1][j][1]
                        sum_b += image[i+1][j][2]
                        count += 1
                    
                    if j > 0 and not (image[i][j-1][0] == 0 and image[i][j-1][1] == 0 and image[i][j-1][2] == 0):
                        sum_r += image[i][j-1][0]
                        sum_g += image[i][j-1][1]
                        sum_b += image[i][j-1][2]
                        count += 1
                    
                    if j < m-1 and not (image[i][j+1][0] == 0 and image[i][j+1][1] == 0 and image[i][j+1][2] == 0):
                        sum_r += image[i][j+1][0]
                        sum_g += image[i][j+1][1]
                        sum_b += image[i][j+1][2]
                        count += 1
                    
                    # 如果有可用相邻点，进行修复
                    if count > 0:
                        temp[i][j][0] = int(math.ceil(sum_r / count))
                        temp[i][j][1] = int(math.ceil(sum_g / count))
                        temp[i][j][2] = int(math.ceil(sum_b / count))
                        changed = True
        
        image = temp  # 更新图片
    
    # 输出结果
    for i in range(n):
        line_parts = []
        for j in range(m):
            line_parts.append(f"R {image[i][j][0]} G {image[i][j][1]} B {image[i][j][2]}")
        print(" ".join(line_parts))

if __name__ == "__main__":
    main()

File: 大一上/图灵班/程序设计荣誉课程-1/yoj全解/YOJ-87/yoj-87.cpp
/*
 * @Author: zmlad
 * @Date: 2025-11-27 18:49:43
 */
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

// 像素结构体
struct Pixel {
    int r, g, b;
};

int main() {
    int n, m;
    cin >> n >> m;
    
    vector<vector<Pixel>> image(n, vector<Pixel>(m));
    
    // 读取输入图片
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            char discard;
            cin >> discard >> image[i][j].r >> discard >> image[i][j].g >> discard >> image[i][j].b;
        }
    }
    
    // 修复过程
    bool changed;
    do {
        changed = false;
        vector<vector<Pixel>> temp = image; // 复制当前状态
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                // 如果是坏点
                if (image[i][j].r == 0 && image[i][j].g == 0 && image[i][j].b == 0) {
                    int sum_r = 0, sum_g = 0, sum_b = 0;
                    int count = 0;
                    
                    // 检查四个方向的相邻点
                    if (i > 0 && !(image[i-1][j].r == 0 && image[i-1][j].g == 0 && image[i-1][j].b == 0)) {
                        sum_r += image[i-1][j].r;
                        sum_g += image[i-1][j].g;
                        sum_b += image[i-1][j].b;
                        count++;
                    }
                    if (i < n-1 && !(image[i+1][j].r == 0 && image[i+1][j].g == 0 && image[i+1][j].b == 0)) {
                        sum_r += image[i+1][j].r;
                        sum_g += image[i+1][j].g;
                        sum_b += image[i+1][j].b;
                        count++;
                    }
                    if (j > 0 && !(image[i][j-1].r == 0 && image[i][j-1].g == 0 && image[i][j-1].b == 0)) {
                        sum_r += image[i][j-1].r;
                        sum_g += image[i][j-1].g;
                        sum_b += image[i][j-1].b;
                        count++;
                    }
                    if (j < m-1 && !(image[i][j+1].r == 0 && image[i][j+1].g == 0 && image[i][j+1].b == 0)) {
                        sum_r += image[i][j+1].r;
                        sum_g += image[i][j+1].g;
                        sum_b += image[i][j+1].b;
                        count++;
                    }
                    
                    // 如果有可用相邻点，进行修复
                    if (count > 0) {
                        temp[i][j].r = ceil(static_cast<double>(sum_r) / count);
                        temp[i][j].g = ceil(static_cast<double>(sum_g) / count);
                        temp[i][j].b = ceil(static_cast<double>(sum_b) / count);
                        changed = true;
                    }
                }
            }
        }
        
        image = temp; // 更新图片
        
    } while (changed);
    
    // 输出结果
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cout << "R " << image[i][j].r << " G " << image[i][j].g << " B " << image[i][j].b;
            if (j < m-1) cout << " ";
        }
        cout << endl;
    }
    
    return 0;
}